File: revolut/ExchangeAccountPickerViewCell.m
//
//  CurrencyView.m
//  revolut
//
//  Created by Alexander Danilov on 02/09/2017.
//  Copyright © 2017 Home. All rights reserved.
//

#import "ExchangeAccountPickerViewCell.h"
#import "AppDelegate.h"
#import "Account.h"

static NSNumberFormatter *currencyFormatter;
static NSNumberFormatter *decimalFormatter;
static NSRegularExpression *removeLeadingZerosRegex;

@implementation ExchangeAccountPickerViewCell {
    AccountManager *_accountManager;
    CurrencyManager *_currencyManager;
    
    UILabel *_fieldLeftView;
    UITapGestureRecognizer *_tapGesture;

    NSString *_accountsKeyPath;
    NSString *_ratesKeyPath;
}

+(void)initialize {
    if (!currencyFormatter) {
        currencyFormatter = [[NSNumberFormatter alloc] init];
        [currencyFormatter setNumberStyle:NSNumberFormatterCurrencyStyle];
        
        decimalFormatter = [[NSNumberFormatter alloc] init];
        decimalFormatter.minimumFractionDigits = 0;
        decimalFormatter.maximumFractionDigits = 2;
        decimalFormatter.minimumIntegerDigits = 1;
        decimalFormatter.currencySymbol = @"";
        
        removeLeadingZerosRegex = [NSRegularExpression
                                   regularExpressionWithPattern:@"^(0*)(\\d.*)"
                                   options:NSRegularExpressionCaseInsensitive
                                   error:nil];
    }
}

-(void)awakeFromNib {
    [super awakeFromNib];
    
    _accountManager = [AppDelegate sharedInstance].accountManager;
    _currencyManager = [AppDelegate sharedInstance].currencyManager;
    
    _fieldLeftView = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, 25, _transferField.frame.size.height)];
    _fieldLeftView.textColor = _transferField.textColor;
    _fieldLeftView.font = _transferField.font;
    
    _transferField.leftView = _fieldLeftView;
    _transferField.leftViewMode = UITextFieldViewModeAlways;
    _transferField.delegate = self;
    
    _tapGesture = [[UITapGestureRecognizer alloc] initWithTarget:_transferField action:@selector(becomeFirstResponder)];
    [self addGestureRecognizer:_tapGesture];
}

-(void)setViewModel:(ExchangeViewModel *)viewModel andType:(AccountPickerType)type andCurrency:(NSString *)currency {
    [self setSubscribed:FALSE];
    
    _viewModel = viewModel;
    _currency = currency;
    _type = type;
    
    _accountsKeyPath = [NSString stringWithFormat:@"accounts.%@.balance", _currency];
    _ratesKeyPath = [NSString stringWithFormat:@"rates.%@", _currency];
    
    if (self.window) {
        [self setSubscribed:TRUE];
    }
    
    _currencyLabel.text = _currency;
    _fieldLeftView.text = _type == kAccountPickerFrom ? @"-" : @"+";
}

-(void)setSubscribed:(BOOL)subscibed {
    if (subscibed) {
        if (_viewModel) {
            [_viewModel addObserver:self forKeyPath:@"valueFrom" options:NSKeyValueObservingOptionNew context:nil];
            [_viewModel addObserver:self forKeyPath:@"currencyFrom" options:NSKeyValueObservingOptionNew context:nil];
            [_viewModel addObserver:self forKeyPath:@"currencyTo" options:NSKeyValueObservingOptionNew context:nil];
        }
        
        if (_currency) {
            NSLog(@"+= <%lu>: %@", (unsigned long)self.hash, _accountsKeyPath);
            [_accountManager addObserver:self forKeyPath:_accountsKeyPath options:NSKeyValueObservingOptionNew|NSKeyValueObservingOptionInitial context:nil];
            [_currencyManager addObserver:self forKeyPath:_ratesKeyPath options:NSKeyValueObservingOptionNew|NSKeyValueObservingOptionInitial context:nil];
        }
    } else {
        if (_viewModel) {
            [_viewModel removeObserver:self forKeyPath:@"valueFrom"];
            [_viewModel removeObserver:self forKeyPath:@"currencyFrom"];
            [_viewModel removeObserver:self forKeyPath:@"currencyTo"];
        }
        
        if (_currency) {
            NSLog(@"-= <%lu>: %@", (unsigned long)self.hash, _accountsKeyPath);
            [_accountManager removeObserver:self forKeyPath:_accountsKeyPath];
            [_currencyManager removeObserver:self forKeyPath:_ratesKeyPath];
        }
    }
}

#pragma mark - UITextField

-(BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string {
    NSCharacterSet *numbersOnly = [NSCharacterSet characterSetWithCharactersInString:@"0123456789,."];
    NSCharacterSet *characterSetFromTextField = [NSCharacterSet characterSetWithCharactersInString:string];
    
    BOOL isNumber = [numbersOnly isSupersetOfSet:characterSetFromTextField];
    if (isNumber) {
        NSString *finalString = [textField.text stringByReplacingCharactersInRange:range withString:string];
        NSDecimalNumber *number = [NSDecimalNumber decimalNumberWithString:finalString];
        if ([number compare:NSDecimalNumber.notANumber] == NSOrderedSame) {
            number = NSDecimalNumber.zero;
            finalString = @"0";
        }
        
        finalString = [removeLeadingZerosRegex stringByReplacingMatchesInString:finalString
                                                                        options:0
                                                                          range:NSMakeRange(0, [finalString length])
                                                                   withTemplate:@"$2"];
        
        textField.text = finalString;
        
        if (_type == kAccountPickerFrom) {
            _viewModel.valueFrom = number;
        } else {
            _viewModel.valueTo = number;
        }
    }
    
    return FALSE;
}

-(void)textFieldDidEndEditing:(UITextField *)textField {
    textField.text = [decimalFormatter stringFromNumber: _type == kAccountPickerFrom ? _viewModel.valueFrom : _viewModel.valueTo];
}

#pragma mark - UIView

-(void)didMoveToWindow {
    [self setSubscribed:!!self.window];
}

#pragma mark - NSObject

-(void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context
{
    if ([keyPath isEqualToString:_accountsKeyPath]) {
        [currencyFormatter setCurrencyCode:_currency];
        Account *account = [_accountManager.accounts valueForKey:_currency];
        _accountValueLabel.text = [NSString localizedStringWithFormat:@"You have %@", [currencyFormatter stringFromNumber:account.balance]];
    }
    
    else if ([keyPath isEqualToString:_ratesKeyPath] || [keyPath isEqualToString:@"valueFrom"] || [keyPath isEqualToString:@"currencyFrom"] || [keyPath isEqualToString:@"currencyTo"]) {
        [currencyFormatter setCurrencyCode:_currency];
        NSDecimalNumber *fromValue = [[NSDecimalNumber alloc] initWithInt:1];
        NSString *fromText = [currencyFormatter stringFromNumber: fromValue];
        
        NSString *currencyTo = _type == kAccountPickerFrom ? _viewModel.currencyTo : _viewModel.currencyFrom;
        [currencyFormatter setCurrencyCode:currencyTo];
        NSDecimalNumber *toValue = [_currencyManager getExchangeRateOfValue:fromValue from:_currency to:currencyTo];
        NSString *toText = [currencyFormatter stringFromNumber: toValue];
        _rateLabel.text = [NSString stringWithFormat:@"%@ = %@", fromText, toText];
        
        NSDecimalNumber *currentExchangeValue = _type == kAccountPickerFrom ? _viewModel.valueFrom : _viewModel.valueTo;
        if ([[NSDecimalNumber decimalNumberWithString:_transferField.text] compare:currentExchangeValue] != NSOrderedSame) {
            _transferField.text = [decimalFormatter stringFromNumber:currentExchangeValue];
        }
    }
    
    else {
        [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
    }
}

@end

File: revolut/ExchangeAccountPickerViewCell.swift
//
//  ExchangeAccountPickerViewCell.swift
//  revolut
//
//  Created by Alexander Danilov on 12/08/2018.
//  Copyright © 2018 Home. All rights reserved.
//

import UIKit
import Foundation

//
//  CurrencyView.m
//  revolut
//
//  Created by Alexander Danilov on 02/09/2017.
//  Copyright © 2017 Home. All rights reserved.
//
class ExchangeAccountPickerViewCell: UICollectionViewCell, UITextFieldDelegate {
    static var currencyFormatter: NumberFormatter = {
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        return formatter
    }()
    
    static let decimalFormatter: NumberFormatter = {
        let formatter = NumberFormatter()
        formatter.minimumFractionDigits = 0
        formatter.maximumFractionDigits = 2
        formatter.minimumIntegerDigits = 1
        formatter.currencySymbol = ""
        return formatter
    }()
    
    static let removeLeadingZerosRegex = {
        return try! NSRegularExpression(pattern: "^(0*)(\\d.*)", options: .caseInsensitive)
    }()
    
    @IBOutlet weak var currencyLabel: UILabel!
    @IBOutlet weak var accountValueLabel: UILabel!
    @IBOutlet weak var rateLabel: UILabel!
    @IBOutlet weak var transferField: UITextField!
    private var fieldLeftView: UILabel!
    private var tapGesture: UITapGestureRecognizer!
    
    private let accountManager: AccountManager
    private let currencyManager: CurrencyManager
    
    var currency: String?
    var viewModel: ExchangeViewModel?
    var type: AccountPickerType?
    private var accountsKeyPath: String?
    private var ratesKeyPath: String?
    
    required init?(coder aDecoder: NSCoder) {
        self.accountManager = AppDelegate.shared.accountManager
        self.currencyManager = AppDelegate.shared.currencyManager
        
        super.init(coder: aDecoder)
    }
    
    override func awakeFromNib() {
        super.awakeFromNib()
        
        self.fieldLeftView = UILabel(frame: CGRect(x: 0, y: 0, width: 25, height: self.transferField.frame.size.height))
        self.fieldLeftView.textColor = self.transferField.textColor
        self.fieldLeftView.font = self.transferField.font
        
        self.transferField.leftView = self.fieldLeftView
        self.transferField.leftViewMode = .always
        self.transferField.delegate = self
        
        self.tapGesture = UITapGestureRecognizer(target: self.transferField, action: #selector(self.becomeFirstResponder))
        self.addGestureRecognizer(tapGesture)
    }
    
    func setViewModel(_ viewModel: ExchangeViewModel, withType type: AccountPickerType, andCurrency currency: String) {
        self.setSubscribed(false)
        
        self.viewModel = viewModel
        self.type = type
        self.currency = currency
        
        self.accountsKeyPath = "accounts.\(currency).balance"
        self.ratesKeyPath = "rates.\(currency)"
        
        if (self.window != nil) {
            self.setSubscribed(true)
        }
    
        self.currencyLabel.text = currency
        self.fieldLeftView.text = type == .from ? "-" : "+"
    }
    
    func setSubscribed(_ subscribed: Bool) {
        if (subscribed) {
            if let viewModel = self.viewModel {
                viewModel.addObserver(self, forKeyPath:"valueFrom", options:.new, context:nil)
                viewModel.addObserver(self, forKeyPath:"currencyFrom", options:.new, context:nil)
                viewModel.addObserver(self, forKeyPath:"currencyTo", options:.new, context:nil)
            }
    
            if self.currency != nil {
                self.accountManager.addObserver(self, forKeyPath:accountsKeyPath!, options:[.new, .initial], context:nil)
                self.currencyManager.addObserver(self, forKeyPath:ratesKeyPath!, options:[.new, .initial], context:nil)
            }
        } else {
            if let viewModel = viewModel {
                viewModel.removeObserver(self, forKeyPath: "valueFrom")
                viewModel.removeObserver(self, forKeyPath: "currencyFrom")
                viewModel.removeObserver(self, forKeyPath: "currencyTo")
            }
    
            if self.currency != nil {
                self.accountManager.removeObserver(self, forKeyPath:self.accountsKeyPath!)
                self.currencyManager.removeObserver(self, forKeyPath:self.ratesKeyPath!)
            }
        }
    }

    // MARK: - UITextField

    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
        let numbersOnly = CharacterSet(charactersIn: "0123456789,.")
        let characterSetFromTextField = CharacterSet(charactersIn: string)
        let isNumber = numbersOnly.isSuperset(of: characterSetFromTextField)
        if isNumber {
            var finalString = (textField.text as NSString?)?.replacingCharacters(in: range, with: string) ?? ""
            var number = NSDecimalNumber(string: finalString)
            if number == NSDecimalNumber.notANumber {
                number = NSDecimalNumber.zero
                finalString = "0"
            }
            
            finalString = ExchangeAccountPickerViewCell.removeLeadingZerosRegex.stringByReplacingMatches(
                in: finalString,
                options: [],
                range: NSRange(location: 0, length: finalString.count),
                withTemplate: "$2")
        
            textField.text = finalString
            
            if type == .from {
                self.viewModel!.valueFrom = number
            } else {
                self.viewModel!.valueTo = number
            }
        }
        
        return false
    }

    func textFieldDidEndEditing(_ textField: UITextField) {
        textField.text = ExchangeAccountPickerViewCell.decimalFormatter.string(from: type == .from ? self.viewModel!.valueFrom : self.viewModel!.valueTo)
    }
    
    // MARK: - UIView
    
    override func didMoveToWindow() {
        super.didMoveToWindow()
        self.setSubscribed(self.window != nil)
    }
    
    // MARK: - NSObject
    
    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        if (keyPath == accountsKeyPath) {
            ExchangeAccountPickerViewCell.currencyFormatter.currencyCode = currency
            let account: Account = self.accountManager.accounts[currency!]! as! Account
            accountValueLabel.text = String.localizedStringWithFormat("You have %@", ExchangeAccountPickerViewCell.currencyFormatter.string(from: account.balance)!)
        } else if (keyPath == ratesKeyPath) || (keyPath == "valueFrom") || (keyPath == "currencyFrom") || (keyPath == "currencyTo") {
            ExchangeAccountPickerViewCell.currencyFormatter.currencyCode = currency
            let fromValue = NSDecimalNumber(value: 1)
            let fromText = ExchangeAccountPickerViewCell.currencyFormatter.string(from: fromValue) ?? ""
            
            let currencyTo: String = type == .from ? viewModel!.currencyTo : viewModel!.currencyFrom
            ExchangeAccountPickerViewCell.currencyFormatter.currencyCode = currencyTo
            let toValue = currencyManager.getExchangeRate(ofValue: fromValue, from: self.currency!, to: currencyTo) ?? NSDecimalNumber.notANumber
            let toText = ExchangeAccountPickerViewCell.currencyFormatter.string(from: toValue) ?? ""
            
            rateLabel.text = "\(fromText) = \(toText)"
            let currentExchangeValue: NSDecimalNumber = type == .from ? viewModel!.valueFrom : viewModel!.valueTo
            if NSDecimalNumber(string: transferField.text) != currentExchangeValue {
                transferField.text = ExchangeAccountPickerViewCell.decimalFormatter.string(from: currentExchangeValue)
            }
        } else {
            super.observeValue(forKeyPath: keyPath, of: object, change: change, context: context)
        }
    }
}
